provider/provider.go

```diff
@@ -0,0 +1,1370 @@
+					next = entry
+					nextKeyFound = true
+				}
+				s.reprovideQueue.Enqueue(entry.Key)
+			}
+			if !nextKeyFound {
+				next = scheduleEntries[0]
+			}
+			// Don't reprovide any region now, but schedule the next one. All regions
+			// are expected to be reprovided when the provider is catching up with
+			// failed regions.
+			s.scheduleNextReprovideNoLock(next.Key, s.timeUntil(next.Data))
+			s.scheduleLk.Unlock()
+			return
+		} else if timeSinceTimerUntilNext < timeSinceTimerRunning {
+			// next is scheduled in the past. While next is in the past, add next to
+			// failedRegions and take nextLeaf as next.
+			count := 0
+			scheduleSize := s.schedule.Size()
+			for timeSinceTimerUntilNext < timeSinceTimerRunning && count < scheduleSize {
+				prefix := next.Key
+				s.reprovideQueue.Enqueue(prefix)
+				next = keyspace.NextNonEmptyLeaf(s.schedule, next.Key, s.order)
+				timeSinceTimerUntilNext = s.timeBetween(s.timeOffset(s.scheduleTimerStartedAt), next.Data)
+				count++
+			}
+		}
+
+		// next is in the future
+		nextPrefix = next.Key
+		timeUntilNextReprovide = s.timeUntil(next.Data)
+	}
+
+	s.scheduleNextReprovideNoLock(nextPrefix, timeUntilNextReprovide)
+	s.scheduleLk.Unlock()
+
+	// If we are offline, don't try to reprovide region.
+	if !s.connectivity.IsOnline() {
+		s.reprovideQueue.Enqueue(currentPrefix)
+		return
+	}
+
+	// Remove prefix that is about to be reprovided from the reprovide queue if
+	// present.
+	s.reprovideQueue.Remove(currentPrefix)
+
+	go func() {
+		s.workerPool.Acquire(periodicWorker)
+		defer s.workerPool.Release(periodicWorker)
+		s.batchReprovide(currentPrefix, true)
+	}()
+}
+
+// scheduleNextReprovideNoLock makes sure the scheduler wakes up in
+// `timeUntilReprovide` to reprovide the region identified by `prefix`.
+func (s *SweepingProvider) scheduleNextReprovideNoLock(prefix bitstr.Key, timeUntilReprovide time.Duration) {
+	s.scheduleCursor = prefix
+	s.scheduleTimer.Reset(timeUntilReprovide)
+	s.scheduleTimerStartedAt = s.clock.Now()
+}
+
+const individualProvideThreshold = 2
+
+func (s *SweepingProvider) failedReprovide(prefix bitstr.Key, err error) {
+	logger.Error(err)
+	// Put prefix in the reprovide queue.
+	s.reprovideQueue.Enqueue(prefix)
+
+	s.connectivity.TriggerCheck()
+}
+
+func (s *SweepingProvider) batchReprovide(prefix bitstr.Key, periodicReprovide bool) {
+	addrInfo, ok := s.selfAddrInfo()
+	if !ok {
+		// Don't provide if the node doesn't have a valid address to include in the
+		// provider record.
+		return
+	}
+
+	// Load keys matching prefix from the keystore.
+	keys, err := s.keyStore.Get(context.Background(), prefix)
+	if err != nil {
+		s.failedReprovide(prefix, fmt.Errorf("couldn't reprovide, error when loading keys: %s", err))
+		if periodicReprovide {
+			s.reschedulePrefix(prefix)
+		}
+		return
+	} else if len(keys) == 0 {
+		logger.Infof("No keys to reprovide for prefix %s", prefix)
+		return
+	} else if len(keys) <= individualProvideThreshold {
```

Comment:
```suggestion
	}
	if len(keys) == 0 {
		logger.Infof("No keys to reprovide for prefix %s", prefix)
		return
	} 
	if len(keys) <= individualProvideThreshold {
```

Reply:
This bit is out of sync with https://github.com/libp2p/go-libp2p-kad-dht/pull/1095